Better auto-vj-uri with new APIs

jota2rz opened the pull request:
- New APIs 'popDeck' and 'trackPaused'
- Better 'auto-vj-uri.html', now handled by decks instead of tracks, video element is never removed and source is changed instead, prevents lags/freezing.
- 'filename' API sanity check when empty.

jota2rz committed:
Better auto-vj-uri with new APIs

- New APIs 'popDeck' and 'trackPaused'
- Better 'auto-vj-uri.html', now handled by decks instead of tracks, video element is never removed and source is changed instead, prevents lags/freezing.
- 'filename' API sanity check when empty.

jota2rz commented:
It should be okay to add jQuery as it's MIT.
I believe a bug with the jQuery transition "fadeTo" may occur and blink twice when moving the crossfader too fast to the left and right but I haven't encountered with it yet.

vladkorotnev commented on assets/3rdparty/js/jquery-3.7.1.min.js:
For a single CSS effect, I would say it's really worth it to just implement it locally, might be even more reliable

vladkorotnev commented on assets/api/api.js:
```diff
@@ -51,20 +51,30 @@ function getFilename(meta) {
             goneTracks.forEach(element => popTrack(element));
         }
 
+        if(typeof popDeck == "function") {
+            let goneDecks = oldDecksId.filter(deck => newDecksId.indexOf(deck) == -1).map(x => decks[x]);
+            goneDecks.forEach(element => popDeck(element));
+        }
```
I would say a `pushDeck` counterpart could be useful for consistency

vladkorotnev commented on assets/auto-vj-uri.html:
```diff
@@ -27,38 +35,45 @@
-                        vLayer.didStart = true;
-                    }
-                });
                 vLayer.addEventListener('ended',() => {
-                    vLayer.style.display = "none";
+                    // Hide video player if it reaches the end to avoid black screen
+                    $(vLayer).fadeOut('fast');
```
If this is the only use of the effect, how about using the CSS animation and opacity properties on the video element? And adding/removing a `.hidden` class for instance.

vladkorotnev commented on assets/api/api.js:
```diff
@@ -77,7 +87,13 @@ function processUpdates(info) {
             }
         }
 
+        if(typeof trackPaused == "function") {
+            let pausedTracks = info.songsOnAir.filter(x => x.isPlaying == false);
+            pausedTracks.forEach(element => trackPaused(element));
+        }
```
Hmm, are you getting tick events for paused decks as well? I don't think my setup does this...

jota2rz replied:
> Hmm, are you getting tick events for paused decks as well? I don't think my setup does this...

I forgot to push the change in logic for this to work! I was supposed to ask if removing the `.is_playing` filter at [logic.rs](https://github.com/vladkorotnev/traktor-obs-relay/blob/204529558d8300204d1b02aefa867fab74cb7e5a/src/logic.rs#L35C8-L35C8) would break other functions.

vladkorotnev replied:
I guess you could just move it into the code above (`trackTick` handler) on the JS side of things. Would it increase the bandwidth of stuff going into the websocket is a whole other question. I'd presume not by a big margin, but don't quote me on that.

jota2rz replied:
> I guess you could just move it into the code above (`trackTick` handler) on the JS side of things. Would it increase the bandwidth of stuff going into the websocket is a whole other question. I'd presume not by a big margin, but don't quote me on that.

That wouldn't work, info has to be sent from the API, and this is the only way.

vladkorotnev replied:
I meant, move the `is_playing` filter from the Rust side of things into the JS one, which is the only one using it as of now

vladkorotnev commented on assets/auto-vj-uri.html:
```diff
@@ -108,6 +150,19 @@
+                }   
+            }
+
+            // Called periodically when any deck is paused
+            function trackPaused(meta) {
```
Is it called periodically while it's paused, or only when the pause happens? The readme says the latter, but this comment says the former

jota2rz replied:
> Is it called periodically while it's paused, or only when the pause happens? The readme says the latter, but this comment says the former

Called periodically while it's paused, fixing the readme.

vladkorotnev commented on src/logic.rs:
```diff
@@ -142,8 +142,14 @@ pub fn get_deck_assoc_filename(deck_id: &Deck, decks: &HashMap<Deck, DeckStatus>
         let subtitle_filename = file_path.file_stem();
-        debug!("Returning filename {}", subtitle_filename.unwrap().to_os_string().into_string().unwrap());
-        return Some(subtitle_filename.unwrap().to_os_string().into_string().unwrap());
+        if subtitle_filename.is_some() {
+            debug!("Returning filename {}", subtitle_filename.unwrap().to_os_string().into_string().unwrap());
+            return Some(subtitle_filename.unwrap().to_os_string().into_string().unwrap());
+        }
+        else {
+            debug!("Filename not found, returning None");
+            None
+        }
```
Is there a case where `file_stem` returns None here?

Also 
```suggestion
   match subtitle_filename {
        Some(subtitle_filename) => {
            debug!("Returning filename {}", subtitle_filename.unwrap().to_os_string().into_string().unwrap());
            return Some(subtitle_filename.unwrap().to_os_string().into_string().unwrap());
        },
        None => {
            debug!("Filename not found, returning None");
            None
        }
    }
```

jota2rz replied:
> Is there a case where `file_stem` returns None here?
> 
> Also

Same answer for `trackPaused`, requires the logic filter change, if the song is removed from the deck and the crossfader has this deck active in the mix, you may get an empty filepath.

vladkorotnev replied:
For this one it's easier since it's  a new API. However please check that other APIs don't break in a similar fashion. (I assume not since they are supposed to deal with file-not-founds to begin with)

vladkorotnev commented:
Also just remembered it might be worth listing the new widget in the [list page](https://github.com/vladkorotnev/traktor-obs-relay/blob/master/assets/index.html)

jota2rz committed:
Transitions without jQuery, new API, is_playing logic now in js

- Transitions are now handled with CSS3 and jQuery has been removed.
- Logic change for songsOnAir, all decks are now passed through and isPlaying is now filtered at Javascript level, this is needed for trackPaused to work correctly.
- Extra sanity check to see if player exist before checking if it's playing and applying transition animations.
- Documentation updates to reflect changes.

jota2rz commented:
@vladkorotnev Everything should be resolved now.

jota2rz committed:
Pause fixes and less console spam

- Pause immediately when trackPaused, a timeout sometimes causes conflict if the track is played too fast again.
- Comment out `trackTick` and `trackPaused` console debug messages to prevent spam.
- Fix for player detection in `popDeck` and `trackPaused`.

jota2rz committed:
Do not play again when it's paused

- Fix for `trackTick` to not resume playing when a track is not playing.

jota2rz committed:
Prevent play interruption by load

This `play` sometimes is requested too fast before `onload` is fired resulting in:

```Uncaught (in promise) DOMException: The play() request was interrupted by a new load request.```

jota2rz commented:
@vladkorotnev Everything should be resolved now.

jota2rz committed:
Handle z-index for players

- Always bring to the top of z-index the last player that starts playing
- Combine CSS to the bottom

jota2rz commented:
@vladkorotnev Everything should be resolved and also did some fixes and improvements.

vladkorotnev commented:
@jota2rz so far lgtm, but have you tested it with the existing widgets just in case?

jota2rz commented:
> @jota2rz so far lgtm, but have you tested it with the existing widgets just in case?

@vladkorotnev No, sorry, I can't test all of them right now.

jota2rz commented:
@vladkorotnev I've tested the other widgets, everything seems to be working fine.
Additional testing by someone else would be appreciated.

vladkorotnev commented:
Thanks! I'll test more later
